src/parsers/PnmlParser.js

```diff
@@ -1,92 +1,126 @@
+import { parse } from 'fast-xml-parser';
+
 export default class PnmlParser {
 
     constructor (metaFactory) {
         this.metaFactory = metaFactory;
-        this.domParser = new DOMParser();
+        this.elements = null;
     }
 
     parse (pnml) {
-        const dom = this.domParser.parseFromString(pnml, 'application/xml');
-
-        const rootElement = dom.documentElement;
-        if (rootElement.nodeName !== 'pnml'
-            || !rootElement.hasChildNodes()) {
+        this.elements = [];
+
+        const data = parse(pnml, {
+            attrNodeName: '_attributes',
+            attributeNamePrefix: '',
+            ignoreAttributes: false,
+            parseAttributeValue: true,
+            parseNodeValue: true,
+            ignoreNameSpace: false,
+            trimValues: true,
+        });
+
+        if (!data.pnml || !data.pnml.net) {
             throw new Error('Invalid document type.');
         }
 
-        const netElement = rootElement.firstElementChild;
-        if (netElement.nodeName !== 'net'
-            || !netElement.hasChildNodes()) {
-            throw new Error('Empty net.');
+        let title = null;
+        if (data.pnml.net.name && data.pnml.net.name.text) {
+            title = data.pnml.net.name.text;
         }
 
-        const treeWalker = document.createTreeWalker(
-            netElement,
-            NodeFilter.SHOW_ELEMENT
-        );
+        // Net elements can be inside a page element
+        const net = data.pnml.net.page || data.pnml.net;
 
-        const elements = [];
-        let lastElement = null;
-        let title = null;
-        while (treeWalker.nextNode()) {
-            switch (treeWalker.currentNode.nodeName) {
+        Object.keys(net).forEach((elementName) => {
+            switch (elementName) {
                 case 'place':
                 case 'transition':
                 case 'arc':
-                    lastElement = this.createElement(treeWalker.currentNode);
-                    elements.push(lastElement);
-                    break;
-                case 'position':
-                    this.setPosition(treeWalker.currentNode, lastElement);
-                    break;
-                case 'dimension':
-                    this.setDimension(treeWalker.currentNode, lastElement);
-                    break;
-                case 'name':
-                    title = this.getName(treeWalker.currentNode);
+                    this.createElement(net, elementName);
                     break;
                 default:
                     // TODO add more shapes or create generic shape
```

Comment:
Generic shapes could be nice. Every other shape could be placed as a polyfigure in renew, so its basicly the same thing, or something like this. Also I think, that generic shapes are toolspecific, so we have to look how to support them correctly. I think that Renew is currently ignoring all other shapes than places, transitions and connections.

Reply:
Yes polyfigures might be a good way forward. This will need some research in the future.